source/fundamentals/crud/bulk.txt

```diff
@@ -0,0 +1,315 @@
+--------------------------
+
+The ``bulkWrite()`` method accepts a list of bulk operations in the
+format of ``WriteModel`` documents. A ``WriteModel`` represents one of
+the CRUD operations. The method also optionally takes
+``BulkWriteOptions`` as a second parameter to specify the order you want
+the operations to be perform.
+
+The following snippet creates a ``List`` of bulk operations to pass into
+the ``bulkWrite()`` method: 
+
+.. literalinclude:: /includes/fundamentals/code-snippets/BulkWrite.java
+   :language: java
+   :dedent:
+   :start-after: begin bulkOpsList
+   :end-before: end bulkOpsList
+
+The following sections show how to create and use each type of
+``WriteModel`` document. The examples in each section contain the
+following documents in a collection:
+
+.. code-block:: json
+
+   { "_id": 1 }
+   { "_id": 2 }
+
+For additional information on the classes and methods mentioned in this
+section, see the following resources: 
+
+- :java-sync-api:`bulkWrite() <com/mongodb/client/MongoCollection.html#bulkWrite(java.util.List,com.mongodb.client.model.BulkWriteOptions)>` API Documentation
+- :java-core-api:`WriteModel <com/mongodb/client/model/WriteModel.html>` API Documentation
+- :java-core-api:`BulkWriteOptions <com/mongodb/client/model/BulkWriteOptions.html>` API Documentation
+
+Insert Operation
+~~~~~~~~~~~~~~~~
+
+To perform an insert operation, create an ``InsertOneModel`` specifying
+the document you want to insert. To insert multiple documents, you must
+create an ``InsertOneModel`` for each document you want to insert.
+
+Example
+```````
+
+The following example creates an ``InsertOneModel`` for two documents
+where their ``_id`` are "3" and "4": 
+
+.. literalinclude:: /includes/fundamentals/code-snippets/BulkWrite.java
+   :language: java
+   :dedent:
+   :start-after: begin insertDocumentsExample
+   :end-before: end insertDocumentsExample
+
+.. note::
+
+   When performing a ``bulkWrite()``, the ``InsertOneModel`` cannot
+   insert a document with an ``_id`` that already exists in the
+   collection. Instead, the method throws an ``MongoBulkWriteException``.
+
+   The following example tries to insert two documents where the ``_id`` is
+   "1" and "3":
+   
+   .. literalinclude:: /includes/fundamentals/code-snippets/BulkWrite.java
+      :language: java
+      :dedent:
+      :start-after: begin insertExceptionExample
+      :end-before: end insertExceptionExample
+
+   The following shows the output of the code above:
+
+   .. code-block::
+
+      Documents in the collection: 
+      {"_id": 1}
+      {"_id": 2}
+
+      A MongoBulkWriteException occured: A MongoBulkWriteException occured with the following message:  
+      Bulk write operation error on server driver-ref-shard-00-02.pw0q4.mongodb.net:27017. 
+      Write errors: [BulkWriteError{index=0, code=11000, message='E11000 duplicate key error collection: 
+      crudOps.bulkWrite index: _id_ dup key: { _id: 1 }', details={}}].
```

Comment:
suggestion: I like this note and example, but I'd remove the part where you log out the documents in the collection since that operation has nothing to do with the exception.

```suggestion
   .. literalinclude:: /includes/fundamentals/code-snippets/BulkWrite.java
      :language: java
      :dedent:
      :start-after: begin insertExceptionExample
      :end-before: end insertExceptionExample

   The following shows the output of the code above:

   .. code-block:: shell

      A MongoBulkWriteException occured: A MongoBulkWriteException occured with the following message:  
      Bulk write operation error on server driver-ref-shard-00-02.pw0q4.mongodb.net:27017. 
      Write errors: [BulkWriteError{index=0, code=11000, message='E11000 duplicate key error collection: 
      crudOps.bulkWrite index: _id_ dup key: { _id: 1 }', details={}}].
```